File: example/ios/JwtTokenModule.swift
//
//  JwtTokenModule.swift
//  ReactNativeSdkExample
//
//  React Native module to generate JWT tokens
//

import Foundation
import React

@objc(JwtTokenModule)
class JwtTokenModule: NSObject {

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }

  @objc
  func generateJwtToken(
    _ secret: String,
    durationMs: Double,
    email: String?,
    userId: String?,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    do {
      let token = try IterableJwtGenerator.generateToken(
        secret: secret,
        durationMs: Int64(durationMs),
        email: email,
        userId: userId
      )
      resolve(token)
    } catch {
      reject("JWT_GENERATION_ERROR", error.localizedDescription, error)
    }
  }
}

File: example/android/app/src/main/java/iterable/reactnativesdk/example/JwtTokenModule.kt
package iterable.reactnativesdk.example

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.Promise
import com.iterable.IterableJwtGenerator
import java.time.Duration

class JwtTokenModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String {
        return NAME
    }

    @ReactMethod
    fun generateJwtToken(
        secret: String,
        durationMs: Double,
        email: String?,
        userId: String?,
        promise: Promise
    ) {
        try {
            val duration = Duration.ofMillis(durationMs.toLong())
            val token = IterableJwtGenerator.generateToken(secret, duration, email, userId)
            promise.resolve(token)
        } catch (e: Exception) {
            promise.reject("JWT_GENERATION_ERROR", e.message, e)
        }
    }

    companion object {
        const val NAME = "JwtTokenModule"
    }
}

